crates/sncast/src/state.rs

```diff
@@ -0,0 +1,324 @@
+pub struct ScriptTransactionEntries {
+    pub transactions: HashMap<String, ScriptTransactionEntry>,
+}
+
+impl ScriptTransactionEntries {
+    #[must_use]
+    pub fn get(&self, key: &str) -> Option<&ScriptTransactionEntry> {
+        self.transactions.get(key)
+    }
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct ScriptTransactionEntry {
+    pub name: String,
+    pub output: ScriptTransactionOutput,
+    pub status: ScriptTransactionStatus,
+    pub timestamp: u32,
+    pub misc: Option<HashMap<String, Value>>,
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+#[serde(tag = "type")]
+pub enum ScriptTransactionOutput {
+    InvokeResponse(InvokeResponse),
+    DeclareResponse(DeclareResponse),
+    DeployResponse(DeployResponse),
+    ErrorResponse(ErrorResponse),
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct ErrorResponse {
+    pub message: String,
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub enum ScriptTransactionStatus {
+    // script executed successfully, transaction accepted/succeeded
+    Success,
+    // script executed successfully, transaction not rejected/reverted
+    Fail,
+    // script error
+    Error,
+}
+
+pub fn load_or_create_state_file(path: &Utf8PathBuf) -> Result<ScriptTransactionsSchema> {
+    if path.exists() {
+        let content = fs::read_to_string(path).expect("Failed to read state file");
+        match serde_json::from_str::<ScriptTransactionsSchema>(&content) {
+            Ok(state_file) => {
+                verify_version(state_file.version)?;
+                Ok(state_file)
+            }
+            Err(_) => Err(anyhow!("Failed to parse state file - it may be corrupt")),
+        }
+    } else {
+        let default_state = ScriptTransactionsSchema {
+            version: STATE_FILE_VERSION,
+            transactions: None,
+        };
+        fs::write(
+            path,
+            serde_json::to_string_pretty(&default_state)
+                .expect("Failed to convert ScriptTransactionsSchema to json"),
+        )
+        .expect("Failed to write initial state to state file");
+        Ok(default_state)
+    }
+}
+
+// todo (1233): remove must_use attribute when it's no longer needed
+#[must_use]
+pub fn generate_transaction_entry_with_id(
+    tx_entry: ScriptTransactionEntry,
+) -> ScriptTransactionEntries {
+    // todo (1545): Implement hashing function for unique ids
+    let id = format!("{}-{}", tx_entry.name, tx_entry.timestamp);
+    let transaction = HashMap::from([(id, tx_entry)]);
+    ScriptTransactionEntries {
+        transactions: transaction,
+    }
+}
+
+pub fn write_txs_to_state_file(
+    state_file_path: &Utf8PathBuf,
+    tx_entries: ScriptTransactionEntries,
+) -> Result<()> {
+    let mut state_file = load_or_create_state_file(state_file_path)
+        .unwrap_or_else(|_| panic!("Failed to write to state file {state_file_path}"));
```

Comment:
```suggestion
        .unwrap_or_else(|_| panic!("Failed to load state file {state_file_path}"));
```

Reply:
Hmm, not sure about this tbh. Maybe I should use `with_context` instead, to include the message from load function? This way we will have a cleaner error message, pointing to the exact problem. wdyt?

Reply:
I like it

Reply:
done https://github.com/foundry-rs/starknet-foundry/pull/1722/commits/66f6ffd83e44492fa7ed0afccd78f9103f30f4fd